fast-danksharding/src/fast_danksharding.rs

```diff
@@ -255,12 +259,188 @@ pub fn main_flow() {
+
+    let mut D_host_flat: Vec<ScalarField> = (0..4 * N_ROWS * FLOW_SIZE).map(|_| ScalarField::zero()).collect();
+    D.copy_to(&mut D_host_flat[..]).unwrap();
+    let D_host_flat = D_host_flat.into_iter().map(|x| { Scalar { s: x } }).collect::<Vec<_>>();
+    let D_host = D_host_flat.chunks(2 * M_POINTS).collect::<Vec<_>>();
+    assert_eq!(D_host, get_debug_data_scalars("D.csv", 2 * N_ROWS, 2 * M_POINTS));
+    
+    println!("Branch2 {:0.3?}", br2_time.elapsed());
+    let D_b4rbo = D_host.clone();
+
+    ////////////////////////////////
+    println!("Branch 3");
+    ////////////////////////////////
+    let br3_time = Instant::now();
+
+    //d0 = MUL_row(d[mu], [S]) 1x8192
+    let d0: Vec<_> = (0..2 * N_ROWS)
+        .map(|i| {
+            let mut s = S.clone();
+            multp_vec(&mut s, &D_b4rbo[i], 0);
+            s
+        })
+        .collect();
+    debug_assert_eq!(
+        d0,
+        get_debug_data_points_xy1("d0.csv", 2 * N_ROWS, 2 * M_POINTS)
+    );
+
+    let mut d1 = vec![Point::infinity(); (2 * N_ROWS) * (2 * M_POINTS / l)];
+    let d0: Vec<_> = d0.into_iter().flatten().collect();
+
+    addp_vec(&mut d1, &d0, 2 * N_ROWS, 2 * M_POINTS, l, 0);
+
+    let d1 = split_vec_to_matrix(&d1, 2 * N_ROWS).clone();
+    debug_assert_eq!(
+        d1,
+        get_debug_data_points_xy1("d1.csv", 2 * N_ROWS, 2 * N_ROWS)
+    );
+
+    let mut delta0: Vec<_> = d1.into_iter().flatten().collect();
+    println!("iecntt batch for delta0");
+    //delta0 = ECINTT_row(d1) 1x512
+    iecntt_batch(&mut delta0, 2 * N_ROWS, 0);
+    debug_assert_eq!(
+        delta0,
+        get_debug_data_points_proj_xy1_vec("delta0.csv", 2 * N_ROWS * 2 * N_ROWS)
+    );
+
+    delta0.chunks_mut(2 * N_ROWS).for_each(|delta0_i| {
+        // delta1 = delta0 << 256 1x512
+        let delta1_i = [&delta0_i[N_ROWS..], &vec![Point::infinity(); N_ROWS]].concat();
+        q_.push(delta1_i);
+    });
+
+    let mut delta1: Vec<_> = q_.into_iter().flatten().collect();
+
+    println!("ecntt batch for delta1");
+    //q[mu] = ECNTT_row(delta1) 1x512
+    ecntt_batch(&mut delta1, 2 * N_ROWS, 0);
+
+    let q_ = split_vec_to_matrix(&delta1, 2 * N_ROWS).clone();
+
+    debug_assert_eq!(
+        q_,
+        get_debug_data_points_xy1("q.csv", 2 * N_ROWS, 2 * N_ROWS)
+    );
+
+    println!("final check");
+
+    let P = q_
+        .iter()
+        .map(|row| list_to_reverse_bit_order(&row.clone()))
+        .collect::<Vec<_>>()
+        .to_vec();
+
+    //final assertion
+    println!("Branch3 {:0.3?}", br3_time.elapsed());
+
+    assert_eq!(
+        P,
+        get_debug_data_points_xy1("P.csv", 2 * N_ROWS, 2 * N_ROWS)
+    );
+
+    assert_ne!(P[12][23], Point::zero()); //dummy check
+    println!("success !!!",);
+}
+
 #[cfg(test)]
 mod tests {
-    use super::main_flow;
+    use super::{main_flow, alternate_flow};
 
     #[test]
     fn test_main_flow() {
```

Comment:
not sure current "main_flow" worth keeping if new API is correct and faster

Reply:
We can remove current main_flow if everyone's happy with the new one, just left it for the sake of easier comparison